### PR TITLE
Smart shed_diff.

### PR DESCRIPTION
**Pull request by jmchilton**
- Fix ``shed_diff`` so its exit code actually is non-zero if there are differences.
 - By default - don't report differences for tool shed populates attributes (``toolshed``, ``changeset_revision``). This can be disabled with ``--raw``.
 - Tests for ensure ``--raw`` and smart differencing are working and exit codes are fixed.
 - Update shed test webapp to inject the tool shed populated fields into respective arguments.

Implements #141.

### CODE REVIEW

**Comment by erasche** [comment]
> Fix shed_diff so its exit code actually is non-zero if there are differences.

+1

**Comment by erasche** [comment]
> Update shed test webapp to inject the tool shed populated fields into respective arguments.

can you just `subprocess.check_command(['sh', 'run_shed.sh', '--daemon'])`?

File: planemo/commands/cmd_shed_diff.py [hunk 1]
```diff
@@ -29,6 +29,12 @@
          " this to testtoolshed.",
     default=None,
 )
+@click.option(
+    "--raw",
```
Comment (erasche) [inline planemo/commands/cmd_shed_diff.py]: unused?

File: planemo/commands/cmd_shed_diff.py [hunk 2]
```diff
@@ -53,7 +59,7 @@ def cli(ctx, path, **kwds):
     def diff(realized_repository):
         working = tempfile.mkdtemp(prefix="tool_shed_diff_")
         try:
-            shed.diff_in(ctx, working, realized_repository, **kwds)
+            return shed.diff_in(ctx, working, realized_repository, **kwds)
```
Comment (erasche) [inline planemo/commands/cmd_shed_diff.py]: only return in one place, returning (I'm assuming) an error code, or None. Maybe better to explicitly return None/-1?  Sorry, it's late, tipsy code review is probably not the best ;)

File: planemo/shed/diff.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    return deps_diff and repos_diff
+
+
+def _shed_diff(file_a, file_b, f=sys.stdout):
+    xml_a = ElementTree.parse(file_a).getroot()
+    xml_b = ElementTree.parse(file_b).getroot()
+    _strip_shed_attributes(xml_a)
+    _strip_shed_attributes(xml_b)
+    return diff.diff(xml_a, xml_b, reporter=f.write)
```
Comment (erasche) [inline planemo/shed/diff.py]: +1

File: planemo/xml/diff.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            reporter('x2 has an attribute x1 is missing: %s'
+                     % name)
+            return False
+    if not text_compare(x1.text, x2.text):
+        reporter('text: %r != %r' % (x1.text, x2.text))
+        return False
+    if not text_compare(x1.tail, x2.tail):
+        reporter('tail: %r != %r' % (x1.tail, x2.tail))
+        return False
+    return _compare_children(x1, x2, reporter)
```
Comment (erasche) [inline planemo/xml/diff.py]: yep, used this too. Good choice.

File: tests/repository_dependencies_shed.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<repositories description="A suite of Galaxy tools designed to work with version 1.2 of the SAMtools package.">
+    <repository changeset_revision="cf875cbe2df4" name="data_manager_sam_fasta_index_builder" owner="devteam" toolshed="https://toolshed.g2.bx.psu.edu" />
```
Comment (erasche) [inline tests/repository_dependencies_shed.xml]: maybe test some without hardcoded `changeset_revision`? Not strictly necessary, just being picky since I'm a bottle and a half in :P

**Comment by erasche** [comment]
+1

**Comment by jmchilton** [comment]
> can you just subprocess.check_command(['sh', 'run_shed.sh', '--daemon'])?

That would really increase the runtime of those tests - it takes well less than a second to start the server and run those tests. I also like mocking out the tool shed interface because it really crystalizes exactly the API planemo is consuming.

There is one functional test that actually exercises a real tool shed (https://github.com/galaxyproject/planemo/blob/master/tests/test_shed.py) and certainly it would be cool if that would automatically start a tool shed and stuff but I have not gotten around to it yet. It is a little bit harder than just ``run_tests.sh`` - since it should configure an isolated environment for the test.

File: planemo/commands/cmd_shed_diff.py (hunk 2 above)
Comment (jmchilton) [inline planemo/commands/cmd_shed_diff.py]: It is the last statement in the diff - it is wrapped in a finally not exception handling so this ``return`` is the only place to return from.

File: planemo/commands/cmd_shed_diff.py (hunk 1 above)
Comment (jmchilton) [inline planemo/commands/cmd_shed_diff.py]: It is used inside of ``planemo.shed``.

File: tests/repository_dependencies_shed.xml (hunk 5 above)
Comment (jmchilton) [inline tests/repository_dependencies_shed.xml]: Well I was a bottle and half in when I wrote this so seems appropriate :). This is meant to emulate the shed version - the plain version is in tests/repository_dependencies.xml - I use this to test that ``shed_diff`` filters the required attributes to make sure these files are actually different.

Thanks for the detailed review!

File: planemo/commands/cmd_shed_diff.py (hunk 1 above)
Comment (erasche) [inline planemo/commands/cmd_shed_diff.py]: Whoops. Missed that. Sorry!

**Comment by erasche** [comment]
That all sounds fair. Definitely harkens back to your point about "reason for travis and a reason for buildbot". 

If you'd like long running tests, the IUC jenkins bot is underutilised! :)

**Comment by jmchilton** [comment]
Thanks for the offer - it isn't so much about CI though. I definitely want that tool shed test running on Travis - another minute is not my concern. I think my other comment was about Galaxy and that minute would drawfs in comparision to the hours it would take to run Galaxy's functional tests. What I love is ``PLANEMO_SKIP_GALAXY_TESTS=1 nosetests`` is taking literally a couple seconds on my personal laptop. A hour matters in CI, a minute doesn't IMO - but a minute does matter during development. I love refactoring quickly and with impunity...

**Comment by erasche** [comment]
Gotcha, sounds good. Didn't realise that shed tests ran that quickly. Made a offhand assumption that it'd be 15-30m to really test all the shed related functionality and corner cases. (Especially if testing against shed@master and shed@dev)
